File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Effect/Combat/SpeechBubble.ts
import { FairyGUI, SuperText } from "csharp";
import Color from "../../DataDefine/Color";
import { View_SpeechBubble } from "../../Gen/View/ModTheSpire_Effect";
import UIHelper from "../../UI/UIHelper";
import { AbstractEffect } from "../AbstractEffect";


export class SpeechBubble extends AbstractEffect{
    private Instance : FairyGUI.GComponent;
    private Bg : FairyGUI.GImage;
    private SuperText:SuperText;
    public constructor(private X:number,private Y:number,private Msg:string,private Interval?:number,private FlipX? : boolean) {
        super();
        this.Instance = <FairyGUI.GComponent>View_SpeechBubble.CreateInstance();
        this.Bg = this.Instance.GetChild("Bg").asImage;
        if(FlipX)
            this.Bg.flip = FairyGUI.FlipType.Horizontal;
        this.Instance.SetPosition(X,Y,0);
        let graph = this.Instance.GetChild("Holder").asGraph;
        this.SuperText = UIHelper.CreateSuperText();
        let wrap = new FairyGUI.GoWrapper(this.SuperText.gameObject);
        wrap.SetScale(100,100)
        this.SuperText.color = Color.Gray.UnityColor();
        this.SuperText.text = `<shake>${Msg}</shake>`;
        this.SuperText.lineSpacing = -50;
        this.SuperText.Width = 2;
        graph.SetNativeObject(wrap);
        this.Instance.GetTransition("t0").Play();
    }
    Update() {
        super.Update();
        this.SuperText.alpha = this.Bg.alpha;
        if(this.Duration >= this.Interval)
            this.IsDone = true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Effect/Combat/SpeechBubble.cs
using Febucci.UI;
using ModTheSpire.DataDefine;
using ModTheSpire.ModTheSpire_Effect;
using UnityEngine;

namespace ModTheSpire.Effect.Combat
{
    public class SpeechBubble : AbstractEffect
    {
        private readonly int mInterval;
        private readonly FairyGUI.GComponent mInstance;
        private readonly FairyGUI.GImage mBg ;
        private readonly TextAnimator mSuperText;

        public SpeechBubble(float x, float y, string msg, int interval = 2000, bool flipX = false)
        {
            mInterval = interval / 1000;
            this.mInstance = View_SpeechBubble.CreateInstance().asCom;
            this.mBg = this.mInstance.GetChild("Bg").asImage;
            if(flipX)
                this.mBg.flip = FairyGUI.FlipType.Horizontal;
            this.mInstance.SetPosition(x,y,0);
            var graph = this.mInstance.GetChild("Holder").asGraph;
            this.mSuperText = Resources.Load<GameObject>("TMP").GetComponent<TextAnimator>();
            var wrap = new FairyGUI.GoWrapper(this.mSuperText.gameObject);
            wrap.SetScale(100, 100);
            var textMeshPro = this.mSuperText.tmproText;
            textMeshPro.color = MTSColor.Gray;
            textMeshPro.text = $"<shake>{msg}</shake>";
            textMeshPro.lineSpacing = -50;
            var size = textMeshPro.rectTransform.sizeDelta;
            size.x = 2;
            textMeshPro.rectTransform.sizeDelta = size;
            graph.SetNativeObject(wrap);
            this.mInstance.GetTransition("t0").Play();
        }

        public override void Update()
        {
            base.Update();
            this.mSuperText.tmproText.alpha = this.mBg.alpha;
            if (this.Duration >= mInterval)
            {
                Finish();
            }
        }

        public override void Dispose()
        {
            base.Dispose();
            mInstance.Dispose();
        }
    }
}